catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserver.java

```diff
@@ -0,0 +1,362 @@
+      entry.destroy();
+      entry.getParent().ifPresent(e -> e.removeChild(entry));
+    }
+  }
+
+  /**
+   * Callback to allow successful children to remove themselves from their parent directory and
+   * putting themselves in a state where they have never been committed.
+   *
+   * @param entry The AsyncFileEntry wrapping the file being listened to.
+   * @param success Boolean that shows if the task failed or completed successfully
+   */
+  private void commitDelete(AsyncFileEntry entry, boolean success) {
+    LOGGER.debug("commitDelete({},{}): Starting...", entry.getName(), success);
+    if (success) {
+      entry.getParent().ifPresent(e -> e.removeChild(entry));
+      entry.destroy();
+    }
+    removeFromProcessors(entry);
+  }
+
+  /**
+   * Steps file by file comparing the snapshot state to the current state of the directory being
+   * monitored.
+   *
+   * @param parent The parent directory (Wrapped in a AsyncFileEntry)
+   * @param previous The list of all children of the parent directory (In sorted order)
+   * @param files The list of current files (in sorted order)
+   */
+  private void checkAndNotify(
+          final AsyncFileEntry parent, final List<AsyncFileEntry> previous, final File[] files) {
+    int c = 0;
+    for (final AsyncFileEntry entry : previous) {
+      while (c < files.length && entry.compareToFile(files[c]) > 0) {
+        doCreate(new AsyncFileEntry(parent, files[c]));
+        c++;
+      }
+      if (c < files.length && entry.compareToFile(files[c]) == 0) {
+        //  Has the file been committed before? We may have matched it
+        //  but if it doesn't exist downstream then a create would be appropriate
+        if (entry.isInitCommit()) {
+          doMatch(entry);
+          checkAndNotify(entry, entry.getChildren(), listFiles(files[c]));
+        } else {
+          doCreate(entry);
+        }
+        c++;
+      } else {
+        //  Do Delete
+        if (!entry.checkNetwork()) {
+          //  The file MAY still exist but it's the network that's down.
+          return;
+        }
+        checkAndNotify(entry, entry.getChildren(), FileUtils.EMPTY_FILE_ARRAY);
+        doDelete(entry);
+      }
+    }
+    for (; c < files.length; c++) {
+      doCreate(new AsyncFileEntry(parent, files[c]));
+    }
+  }
+
+  /**
+   * Called when the observer should compare the snapshot state to the actual state of the directory
+   * being monitored.
+   */
+  public void checkAndNotify() {
+
+    if (listener == null) {
+      return;
+    }
+
+    /* fire onStart() */
+    listener.onStart(this);
+
+    /* fire directory/file events */
+    final File root = rootFile.getFile();
+    if (root.exists()) {
+      checkAndNotify(rootFile, rootFile.getChildren(), listFiles(rootFile.getFile()));
+    } else if (rootFile.exists()) {
+      checkAndNotify(rootFile, rootFile.getChildren(), FileUtils.EMPTY_FILE_ARRAY);
+    } else {
+      // Didn't exist and still doesn't
+    } //  rootFile.refresh()
```

Comment:
📘 Kept this commented out but I'm not sure if I want to put it there or call hasChanged on the root file. 
having `if(rootFile.hasChanged()) { rootFile.commit(); }` seemed weird to me.

Adding something like this in would cause less processing if there is a long network disconnect